src/serac/physics/utilities/variational_form/integral.hpp

```diff
@@ -0,0 +1,597 @@
+ * @see mfem::GeometricFactors
+ * @param[in] num_elements The number of elements in the mesh
+ */
+template <Geometry g, typename test, typename trial, int geometry_dim, int spatial_dim, int Q,
+          typename derivatives_type>
+void gradient_kernel(const mfem::Vector& dU, mfem::Vector& dR, derivatives_type* derivatives_ptr,
+                     const mfem::Vector& J_, int num_elements)
+{
+  using test_element               = finite_element<g, test>;
+  using trial_element              = finite_element<g, trial>;
+  using element_residual_type      = typename trial_element::residual_type;
+  static constexpr int  test_ndof  = test_element::ndof;
+  static constexpr int  trial_ndof = trial_element::ndof;
+  static constexpr auto rule       = GaussQuadratureRule<g, Q>();
+
+  // mfem provides this information in 1D arrays, so we reshape it
+  // into strided multidimensional arrays before using
+  auto J  = mfem::Reshape(J_.Read(), rule.size(), spatial_dim, geometry_dim, num_elements);
+  auto du = impl::Reshape<trial>(dU.Read(), trial_ndof, num_elements);
+  auto dr = impl::Reshape<test>(dR.ReadWrite(), test_ndof, num_elements);
+
+  // for each element in the domain
+  for (int e = 0; e < num_elements; e++) {
+    // get the (change in) values for this particular element
+    tensor du_elem = impl::Load<trial_element>(du, e);
+
+    // this is where we will accumulate the (change in) element residual tensor
+    element_residual_type dr_elem{};
+
+    // for each quadrature point in the element
+    for (int q = 0; q < static_cast<int>(rule.size()); q++) {
+      // get the position of this quadrature point in the parent and physical space,
+      // and calculate the measure of that point in physical space.
+      auto   xi  = rule.points[q];
+      auto   dxi = rule.weights[q];
+      auto   J_q = make_tensor<spatial_dim, geometry_dim>([&](int i, int j) { return J(q, i, j, e); });
+      double dx  = impl::Measure(J_q) * dxi;
+
+      // evaluate the (change in) value/derivatives at this quadrature point
+      auto darg = impl::Preprocess<trial_element>(du_elem, xi, J_q);
+
+      // recall the derivative of the q-function w.r.t. its arguments at this quadrature point
+      auto dq_darg = derivatives_ptr[e * int(rule.size()) + q];
+
+      // use the chain rule to compute the first-order change in the q-function output
+      auto dq = chain_rule(dq_darg, darg);
+
+      // integrate dq against test space shape functions / gradients
+      // to get the (change in) element residual contributions
+      dr_elem += impl::Postprocess<test_element>(dq, xi, J_q) * dx;
+    }
+
+    // once we've finished the element integration loop, write our element residuals
+    // out to memory, to be later assembled into global residuals by mfem
+    impl::Add(dr, dr_elem, e);
+  }
+}
+
+namespace impl {
+template <typename spaces>
+struct get_trial_space;  // undefined
```

Comment:
I don't think we need the extra `get`s.

Reply:
I'm not sure what you mean, these are used in the `Integral` class template. Are you asking for it to be refactored to not use them?

Reply:
oh, I just meant to rename them `get_trial_space` -> `trial_space`.

Reply:
I'm going to leave the `get_` prefix since shortening it to `trial_space` can conflict with other things by the same name used in WeakForm, Integral, etc